substrate/frame/preimage/precompiles/src/interface/IPreimage.sol

```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.30;
+
+/// @dev The on-chain address of the Preimage precompile.
+address constant PREIMAGE_PRECOMPILE_ADDRESS = address(0xD0000);
+
+interface IPreimage {
+	/// @dev Register a preimage on-chain.
+	/// @param preImage The preimage to be registered on-chain
+	/// @return hash The hash of the preimage
+	function notePreimage(bytes memory preImage) external returns (bytes32 hash);
+
+	/// @dev Clear an unrequested preimage from storage.
+	/// @param hash The preimage to be cleared from storage
+	function unnotePreimage(bytes32 hash) external;
```

Comment:
Can this error if the preimage does not exist?

Reply:
If a contract tries to unnote a hash that is not its own then it sounds like a bug and will probably be handled differently than the when the preimage does not exist.

But not sure how you usually do these precompiles.